### ISSUE TITLE
Retain data validation annotation attributes in model generation

### ISSUE THREAD

**Issue by cbordeman** [NSwag/NSwag#301]
Attributes such as [Required], [Range], [StringLength(5)] from the System.ComponentModel.DataAnnotations namespace should be retained in the generated client side models so validation can be performed.

**Comment by rsuter** [comment]
Good idea. However this is an NJsonSchema issue, can you create one on http://njsonschema.org?

**Comment by jonnybi** [comment]
:thumbsup: waiting for this too

**Comment by cbordeman** [comment]
I don't see how this is an NJsonSchema issue, since it involves scanning the webapi compiled assembly, which I believe is done by NSwagStudio.

**Comment by rsuter** [comment]
DTO classes and interface are generated by njsonschema...

**Comment by cbordeman** [comment]
NSwagStudio has the option to scan a dll and extract c# attributes, and that code appears to be in NSwagStudio.  So doesn't part of the implementation need to reside in NSwagStudio ?

**Comment by jelical** [comment]
NJsonSchema  includes proper defs for validation.
`"RegistrationViewModel": {
      "type": "object",
      "additionalProperties": false,
      "required": [
        "FirstName",
        "LastName"
      ],
      "properties": {
        "UserName": {
          "type": "string",
          "format": "email",
          "maxLength": 60
        },
        "FirstName": {
          "type": "string",
          "maxLength": 140
        },
        "LastName": {
          "type": "string",
          "maxLength": 140
        },
        "Password": {
          "type": "string"
        }
      },`

But there are no manifestation of this in generated models for TypeScript

**Comment by rsuter** [comment]
There are no commonly available attributes for typescript... what exactly would you expect?

**Comment by jelical** [comment]
Ideally, some kind of optional extensibility mechanism to obtain result like [here](https://github.com/pleerock/class-validator). I am not speaking about integrating with this particular library, but support of some kind of map between NJsonSchema validation metadata attributes and TS decorators.

**Comment by adamlubek** [comment]
it would be good to even have these available as constant values so they can be passed to client side validation mechanism e.g. if type had a property/constant containing value of StringLength attribute, then I could use that in angular reactive form validators. This way I could keep API and client in sync

**Comment by berhir** [comment]
We are also looking for a way to use the validation values from the swagger definition in angular reactive form validators. The solution suggested by @adamlubek to have it available as constant values would work for us.

[swagger-ts-generator](https://github.com/areijngoudt/swagger-ts-generator) creates a FormGroup for each model. But most of the time we don't have a 1:1 mapping of the DTO in the UI, so for us this is not so helpful.

For our scenario it would be nice to have a constant for each field with an array of the generated reactive form validators. But to have at least the values would be a good start.

**Comment by adamlubek** [comment]
apologies for shameless plug but I created npm package for getting typescript validation constants from c# dll, maybe it would help someone on c#/typescript stack. https://www.npmjs.com/package/csharp-data-annotations-parser

**Comment by blasky** [comment]
Has there been any movement on this issue, either on the NSwag or NJsonSchema side of things?...i.e. some sort of exposure of the relevant annotations in the generated models such that some other code can make use of them?

**Comment by cbordeman** [comment]
NJsonSchema already supports standard data validation attributes in the generated schema, as someone indicated.  I don't know if NSwag uses them yet.

**Comment by RSuter** [comment]
NSwag uses NJS for DTOs so this should work. But there is a setting to disable that, maybe it’s switched off?

**Comment by blasky** [comment]
Good news. Thanks for the quick reply. I will investigate and share if I find it.

But are you saying there may be a config setting written out to the *.nswag configuration document (...which appears to drive the swaggerToTypeScriptClient execution)? Or a setting in one of the NJsonSchema xml config documents to tweak?

And any idea what I would be looking for in the resultant TypeScript service code upon success? Do we think this should show up as info Decorators on the TypeScript object class?

**Comment by RSuter** [comment]
Validation and data annotations are only supported in the c# output

**Comment by blasky** [comment]
OK. So this remains an open issue for the TypeScript client code, which currently receives no annotation info from the C# source object. Is this on the road map for NSwag or NJsonSchema.CodeGeneration.TypeScript (...wherever the ultimate solution resides)?

Thanks. Just trying to understand and evaluate the tool and feature set at this time.

**Comment by RicoSuter** [comment]
@blasky currently not planned, moved to NJS.
